File: javascript/exercises/practice/unit_test_80b64036_a033_4eec_8bae_3911526143dc/txn_coordinator/.meta/txn_coordinator.js
const crypto = require('crypto');

function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function callWithTimeout(promise, timeout) {
  return Promise.race([
    promise,
    new Promise((_, reject) => setTimeout(() => reject(new Error('timeout')), timeout))
  ]);
}

async function attemptPrepare(service, transactionId, config) {
  let attempts = config.retryAttempts || 1;
  let backoff = 10;
  let currentAttempt = 0;
  while (currentAttempt < attempts) {
    try {
      if (config.logging) {
        config.logging.log(`Service prepare attempt ${currentAttempt + 1} for transaction ${transactionId}`);
      }
      const result = await callWithTimeout(service.prepare(transactionId), config.timeout || 1000);
      if (result === 'vote commit') {
        return;
      } else {
        throw new Error('vote abort');
      }
    } catch (err) {
      currentAttempt++;
      if (currentAttempt < attempts) {
        if (config.logging) {
          config.logging.log(`Service prepare failed on attempt ${currentAttempt} for transaction ${transactionId}, retrying...`);
        }
        await delay(backoff);
        backoff *= 2;
      } else {
        if (config.logging) {
          config.logging.log(`Service prepare failed after ${attempts} attempts for transaction ${transactionId}`);
        }
        throw err;
      }
    }
  }
}

async function transactionCoordinator(services, config = {}) {
  let transactionId;
  if (crypto.randomUUID) {
    transactionId = crypto.randomUUID();
  } else {
    transactionId = `${Date.now()}_${Math.random().toString(36).substring(2, 10)}`;
  }
  if (config.logging) {
    config.logging.log(`Transaction ${transactionId} started.`);
  }
  try {
    await Promise.all(services.map(service => attemptPrepare(service, transactionId, config)));
  } catch (prepareError) {
    if (config.logging) {
      config.logging.log(`Prepare phase failed for transaction ${transactionId}: ${prepareError.message}`);
    }
    await Promise.all(services.map(service => {
      return service.rollback(transactionId).catch(() => {});
    }));
    if (config.logging) {
      config.logging.log(`Transaction ${transactionId} rolled back.`);
    }
    return false;
  }
  try {
    await Promise.all(services.map(service => service.commit(transactionId)));
  } catch (commitError) {
    if (config.logging) {
      config.logging.log(`Commit phase encountered error for transaction ${transactionId}: ${commitError.message}`);
    }
    await Promise.all(services.map(service => {
      return service.rollback(transactionId).catch(() => {});
    }));
    if (config.logging) {
      config.logging.log(`Transaction ${transactionId} rolled back due to commit failure.`);
    }
    return false;
  }
  if (config.logging) {
    config.logging.log(`Transaction ${transactionId} committed successfully.`);
  }
  return true;
}

module.exports = { transactionCoordinator };

File: go/exercises/practice/unit_test_06b98bec_968c_4ae6_82c3_4eee4a3f8c62/txn_coordinator/.meta/txn_coordinator.go
package txncoordinator

import (
	"errors"
	"fmt"
	"log"
	"sync"
	"time"
)

// Constants for configuration
const (
	defaultTimeout     = 5 * time.Second
	maxRetries        = 3
	baseRetryInterval = 100 * time.Millisecond
)

// Participant interface defines the contract for participating services
type Participant interface {
	Prepare(transactionID string, data map[string]interface{}) error
	Commit(transactionID string) error
	Rollback(transactionID string) error
}

// Transaction represents a distributed transaction
type Transaction struct {
	ID           string
	Participants []Participant
	Status       TransactionStatus
	mu           sync.RWMutex
}

// TransactionStatus represents the current state of a transaction
type TransactionStatus int

const (
	StatusActive TransactionStatus = iota
	StatusPreparing
	StatusCommitting
	StatusRollingBack
	StatusCommitted
	StatusRolledBack
)

// Coordinator manages distributed transactions
type Coordinator struct {
	transactions map[string]*Transaction
	mu           sync.RWMutex
	idCounter    uint64
}

// NewCoordinator creates a new transaction coordinator
func NewCoordinator() *Coordinator {
	return &Coordinator{
		transactions: make(map[string]*Transaction),
	}
}

// BeginTransaction starts a new transaction
func (c *Coordinator) BeginTransaction() string {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.idCounter++
	txID := fmt.Sprintf("tx-%d", c.idCounter)
	
	transaction := &Transaction{
		ID:     txID,
		Status: StatusActive,
	}
	
	c.transactions[txID] = transaction
	log.Printf("Transaction started: %s", txID)
	return txID
}

// EnlistParticipant adds a participant to an existing transaction
func (c *Coordinator) EnlistParticipant(transactionID string, participant Participant) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	tx, exists := c.transactions[transactionID]
	if !exists {
		return fmt.Errorf("transaction %s not found", transactionID)
	}

	tx.mu.Lock()
	defer tx.mu.Unlock()

	if tx.Status != StatusActive {
		return fmt.Errorf("transaction %s is not active", transactionID)
	}

	tx.Participants = append(tx.Participants, participant)
	log.Printf("Participant enlisted in transaction %s", transactionID)
	return nil
}

// CommitTransaction attempts to commit the transaction using 2PC
func (c *Coordinator) CommitTransaction(transactionID string, data map[string]interface{}) error {
	c.mu.RLock()
	tx, exists := c.transactions[transactionID]
	c.mu.RUnlock()

	if !exists {
		return fmt.Errorf("transaction %s not found", transactionID)
	}

	tx.mu.Lock()
	if tx.Status != StatusActive {
		tx.mu.Unlock()
		return fmt.Errorf("transaction %s is not active", transactionID)
	}
	tx.Status = StatusPreparing
	tx.mu.Unlock()

	// Prepare phase
	prepareResults := make(chan error, len(tx.Participants))
	var wg sync.WaitGroup

	for _, p := range tx.Participants {
		wg.Add(1)
		go func(participant Participant) {
			defer wg.Done()
			err := c.prepareWithRetry(participant, transactionID, data)
			prepareResults <- err
		}(p)
	}

	// Wait for all prepare operations with timeout
	done := make(chan struct{})
	go func() {
		wg.Wait()
		close(done)
	}()

	select {
	case <-done:
		// Continue with processing results
	case <-time.After(defaultTimeout):
		log.Printf("Prepare phase timeout for transaction %s", transactionID)
		c.rollbackTransaction(tx)
		return errors.New("prepare phase timeout")
	}

	// Check prepare results
	close(prepareResults)
	for err := range prepareResults {
		if err != nil {
			log.Printf("Prepare phase failed for transaction %s: %v", transactionID, err)
			c.rollbackTransaction(tx)
			return fmt.Errorf("prepare phase failed: %v", err)
		}
	}

	// Commit phase
	tx.mu.Lock()
	tx.Status = StatusCommitting
	tx.mu.Unlock()

	var commitError error
	for _, p := range tx.Participants {
		if err := c.commitWithRetry(p, transactionID); err != nil {
			commitError = err
			log.Printf("Commit failed for transaction %s: %v", transactionID, err)
			// Continue committing other participants
		}
	}

	tx.mu.Lock()
	tx.Status = StatusCommitted
	tx.mu.Unlock()

	if commitError != nil {
		return fmt.Errorf("commit phase failed: %v", commitError)
	}

	log.Printf("Transaction %s committed successfully", transactionID)
	return nil
}

// RollbackTransaction explicitly rolls back a transaction
func (c *Coordinator) RollbackTransaction(transactionID string) error {
	c.mu.RLock()
	tx, exists := c.transactions[transactionID]
	c.mu.RUnlock()

	if !exists {
		return fmt.Errorf("transaction %s not found", transactionID)
	}

	return c.rollbackTransaction(tx)
}

// rollbackTransaction performs the rollback operation
func (c *Coordinator) rollbackTransaction(tx *Transaction) error {
	tx.mu.Lock()
	if tx.Status == StatusRolledBack {
		tx.mu.Unlock()
		return nil
	}
	tx.Status = StatusRollingBack
	tx.mu.Unlock()

	var rollbackError error
	for _, p := range tx.Participants {
		if err := c.rollbackWithRetry(p, tx.ID); err != nil {
			rollbackError = err
			log.Printf("Rollback failed for transaction %s: %v", tx.ID, err)
			// Continue rolling back other participants
		}
	}

	tx.mu.Lock()
	tx.Status = StatusRolledBack
	tx.mu.Unlock()

	if rollbackError != nil {
		return fmt.Errorf("rollback failed: %v", rollbackError)
	}

	log.Printf("Transaction %s rolled back successfully", tx.ID)
	return nil
}

// Helper functions for retrying operations with exponential backoff

func (c *Coordinator) prepareWithRetry(p Participant, txID string, data map[string]interface{}) error {
	return c.retryOperation(func() error {
		return p.Prepare(txID, data)
	})
}

func (c *Coordinator) commitWithRetry(p Participant, txID string) error {
	return c.retryOperation(func() error {
		return p.Commit(txID)
	})
}

func (c *Coordinator) rollbackWithRetry(p Participant, txID string) error {
	return c.retryOperation(func() error {
		return p.Rollback(txID)
	})
}

func (c *Coordinator) retryOperation(operation func() error) error {
	var lastErr error
	for attempt := 0; attempt < maxRetries; attempt++ {
		if attempt > 0 {
			backoff := time.Duration(1<<uint(attempt-1)) * baseRetryInterval
			time.Sleep(backoff)
		}

		if err := operation(); err != nil {
			lastErr = err
			log.Printf("Operation failed (attempt %d/%d): %v", attempt+1, maxRetries, err)
			continue
		}
		return nil
	}
	return lastErr
}